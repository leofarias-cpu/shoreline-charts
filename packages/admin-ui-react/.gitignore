# Automatically generated
/cjs
/context
/createIcons
/createSystem
/domElements
/esm
/hooks
/jsx
/symbols
/system
/tag
/themeMode
/ts
/types
/umd
/util
