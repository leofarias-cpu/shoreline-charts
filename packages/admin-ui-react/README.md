# Onda component model

## jsx

You can use the `jsx` function to create new components using elements or components as reference. It generates an `OndaJSXComponent` that inherits and enhances the reference types and styles.

### Referencing elements

To reference elements, simply pass its string representation to the function.

```jsx noInline
const BlueSquare = jsx('div')({
  bg: 'blue',
  size: 50,
})

render(<BlueSquare />)
```

#### Aliasing

`jsx` is aliased with any valid `JSX.Element` for faster typing.

```jsx noInline
const BlueSquare = jsx('div')({
  bg: 'blue',
  size: 50,
})

render(<BlueSquare />)
```

### Referencing components

The idea is the same as elements. It will inherit the types and behavior of the passed component.

```jsx noInline
import { motion } from 'framer-motion'

const Scale = jsx(motion.div)({
  bg: 'blue',
  size: 50,
})

render(<Scale animate={{ scale: 2 }} transition={{ duration: 0.5 }} />)
```

### Variants

`Variants` are props that change the component look. You can set their default values within `defaultProps`, as shown in the example below.

```jsx noInline
const Square = jsx('div')({
  size: 100,
  variants: {
    color: {
      blue: {
        bg: 'blue',
        ':hover': {
          bg: 'blue.hover',
        },
      },
      red: {
        bg: 'red',
        ':hover': {
          bg: 'red.hover',
        },
      },
    },
  },
})

Square.defaultProps = {
  color: 'blue',
}

render(
  <Set>
    <Square />
    <Square color="red" />
  </Set>
)
```

#### Synching variants

In some cases, you may want to attach styles if two or more variants have specific values. This happens when a variant style depends on other variants, and we call this a `variant overlap`. To deal with that you can use the `sync` array, which has the type:

| Name                  | Type                | Description                               | Required | Default |
| --------------------- | ------------------- | ----------------------------------------- | -------- | ------- |
| csx                   | `StyleObject`       | Style to apply if the two variants exists | âœ…       | ðŸš«      |
| [k in keyof variants] | `keyof variants[k]` | Variants values to sync                   | ðŸš«       | ðŸš«      |

In the example below, we have a square with a `solid` or an `outline` filling. The square can also be either `red` or `blue`. We have a `variant overlap` here because to set the final style of our square we need to know the value of both `fill` and `color` variants.

```jsx noInline
const Square = jsx('div')(
  {
    size: 100,
    variants: {
      fill: {
        outline: {
          borderWidth: 2,
          borderStyle: 'solid',
        },
        solid: {
          border: 'none',
        },
      },
      color: {
        blue: {
          bg: 'blue',
          ':hover': {
            bg: 'blue.hover',
          },
        },
        red: {
          bg: 'red',
          ':hover': {
            bg: 'red.hover',
          },
        },
      },
    },
  },
  [
    {
      // if the color is blue and fill outline
      color: 'blue',
      fill: 'outline',
      csx: {
        bg: 'transparent',
        color: 'blue',
        ':hover': {
          bg: 'blue.secondary.hover',
        },
      },
    },
    {
      // if the color is red and fill outline
      color: 'red',
      fill: 'outline',
      csx: {
        bg: 'transparent',
        color: 'red',
        ':hover': {
          bg: 'red.secondary.hover',
        },
      },
    },
  ]
)

Square.defaultProps = {
  color: 'blue',
  fill: 'solid',
}

render(
  <Set>
    <Square />
    <Square fill="outline" />
    <Square color="red" />
    <Square color="red" fill="outline" />
  </Set>
)
```

### Options

Options are used to add extra behavior on the component.

```tsx noInline
interface DisplayOptions {
  visible?: boolean
}

const Display = jsx('div')({}, {
  // you need to declare the used options here.
  options: ['visible'],
  /**
   * options: component options
   * props: other props
   * system: return of admin-core/useSystem
   */
  useOptions(options: DisplayOptions, props, system) {
    const { children, ...divProps } = props
    const { visible } options

    return {
      ...divProps,
      children: visible ? children : null
    }
  }
})

Display.defaultProps = {
  visible: true
}
```

### Polymorphism

Every component generated by `jsx` is polymorphic. With the `as` prop you can define the component render type, to extend or change its behavior.

```jsx noInine
const Button = jsx('button')({
  cursor: 'pointer',
  bg: 'blue',
  color: 'light.primary',
  fontSize: 1,
  height: 32,
  display: 'flex',
  alignItems: 'center',
  justifyContent: 'center',
  paddingX: 2,
  borderRadius: 'default',
  textAlign: 'center',
})

render(
  <Set>
    <Button>Button</Button>
    <Button as="a" href="">
      Button that is a link
    </Button>
  </Set>
)
```

### Typescript Props

You can obtain the type of any component created with `jsx` using:

- `React.ComponentPropsWithRef`: When refs are supported
- `React.ComponentPropsWithoutRef`: When refs are not supported

Example without options:

```ts isStatic
const Button = jsx('button')({})

type ButtonProps = React.ComponentPropsWithRef<typeof Button>
```

Example with options:

```ts isStatic
const Button = jsx('button')(
  {},
  {
    useOptions(options: ButtonOptions, props) {
      /** ... */
    },
  }
)

interface ButtonOptions {}

type ButtonProps = React.ComponentPropsWithRef<typeof Button> & ButtonOptions
```
