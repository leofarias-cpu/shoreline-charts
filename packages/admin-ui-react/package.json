{
  "name": "@vtex/admin-ui-react",
  "license": "MIT",
  "version": "0.9.6-dev.4",
  "main": "dist/vtex-admin-ui-react.cjs.js",
  "module": "dist/vtex-admin-ui-react.esm.js",
  "types": "dist/vtex-admin-ui-react.cjs.d.ts",
  "publishConfig": {
    "access": "public",
    "registry": "https://registry.npmjs.org"
  },
  "files": [
    "dist",
    "src"
  ],
  "engines": {
    "node": ">=10"
  },
  "scripts": {
    "prebuild": "rm -rf dist",
    "dev": "rollup -c -w",
    "build": "npm run prebuild && rollup -c"
  },
  "repository": {
    "directory": "packages/react",
    "type": "git",
    "url": "git+https://github.com/vtex/admin-ui.git"
  },
  "bugs": {
    "url": "https://github.com/vtex/admin-ui/issues"
  },
  "peerDependencies": {
    "@vtex/admin-ui-core": ">=0.x",
    "@vtex/admin-ui-util": ">=0.x",
    "@vtex/phosphor-icons": ">=0.x",
    "react": "^16.9 || ^17.0",
    "react-dom": "^16.9 || ^17.0"
  },
  "devDependencies": {
    "@types/react-helmet": "^6.1.5",
    "@vtex/admin-ui-core": "^0.12.5-dev.4",
    "@vtex/admin-ui-util": "^0.4.4-dev.3",
    "@vtex/phosphor-icons": "0.2.8"
  },
  "dependencies": {
    "focus-visible": "^5.2.0",
    "react-helmet": "^6.1.0",
    "react-is": "^17.0.2",
    "tiny-invariant": "^1.1.0",
    "tiny-warning": "^1.0.3"
  }
}
