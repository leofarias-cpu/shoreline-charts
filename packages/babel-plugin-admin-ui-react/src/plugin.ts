import type * as BabelCore from '@babel/core'
import type { Binding } from '@babel/traverse'

type PluginState = BabelCore.PluginPass & {
  jsxImportNode?: BabelCore.types.ImportSpecifier
  jsxBinding?: Binding
}

export function adminUiPlugin({
  types: t,
}: typeof BabelCore): BabelCore.PluginObj<PluginState> {
  return {
    post() {
      delete this.jsxImportNode
      delete this.jsxBinding
    },
    visitor: {
      ImportDeclaration(path) {
        if (path.node.source.value === '@vtex/admin-ui-react') {
          this.jsxImportNode = path.node.specifiers.find(
            (specifier): specifier is BabelCore.types.ImportSpecifier => {
              return (
                specifier.type === 'ImportSpecifier' &&
                specifier.imported.type === 'Identifier' &&
                specifier.local.name === 'jsx'
              )
            }
          )

          this.jsxBinding = this.file.path.scope.getOwnBinding('jsx')
        }
      },

      CallExpression(path) {
        if (!this.jsxImportNode || !this.jsxBinding) return

        const { node, parentPath } = path

        const isJsxImported = (identifier: BabelCore.types.Expression) => {
          return this.jsxBinding!.referencePaths.some((referencePath) => {
            return referencePath.node === identifier
          })
        }

        if (
          node.callee.type === 'CallExpression' &&
          node.callee.callee.type === 'Identifier' &&
          isJsxImported(node.callee.callee) &&
          parentPath.node.type === 'VariableDeclarator' &&
          parentPath.node.id.type === 'Identifier' &&
          parentPath.parent.type === 'VariableDeclaration'
        ) {
          path.replaceWith(
            t.callExpression(
              t.memberExpression(
                t.identifier('Object'),
                t.identifier('assign')
              ),
              [
                node,
                t.objectExpression([
                  t.objectProperty(
                    t.identifier('displayName'),
                    t.stringLiteral(parentPath.node.id.name)
                  ),
                ]),
              ]
            )
          )
        }
      },
    },
  }
}
