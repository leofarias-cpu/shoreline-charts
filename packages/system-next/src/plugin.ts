import invariant from 'tiny-invariant'

const PLUGIN_NAME = Symbol('plugin name')
const PLUGIN_NAMESPACES = Symbol('plugin namespaces')
const PLUGIN_ENTRIES = Symbol('plugin entries')
const PLUGIN_ALIASES = Symbol('plugin aliases')
const PLUGIN_RULES = Symbol('plugin rules')
const PLUGIN_TRANSFORMS = Symbol('plugin transforms')
const PLUGIN_SPLITS = Symbol('plugin splits')

/**
 * Creates a valid onda plugin
 * @example
 * createPlugin({
 *  name: 'onda-plugin-x',
 *  namespaces: [],
 *  entries: () => {},
 *  aliases: {},
 *  rules: {},
 *  transforms: {},
 *  splits: {}
 * })
 */
export function createPlugin<Theme extends Record<string, any>>(
  params: PluginParams<Theme>
): Plugin<Theme> {
  const { name, namespaces, entries, aliases, rules, transforms, splits } =
    params

  const invariants = getPluginInvariants(params)

  invariant(invariants.name.constraint, invariants.name.message)
  invariant(invariants.rules.constraint, invariants.rules.message)

  return {
    [PLUGIN_NAME]: name,
    [PLUGIN_NAMESPACES]: namespaces,
    [PLUGIN_ENTRIES]: entries ?? {},
    [PLUGIN_ALIASES]: aliases ?? {},
    [PLUGIN_RULES]: rules ?? {},
    [PLUGIN_TRANSFORMS]: transforms ?? {},
    [PLUGIN_SPLITS]: splits ?? {},
  }
}

/**
 * Get plugin invariants
 * @param params
 */
function getPluginInvariants(params: PluginParams<any>) {
  const errorWhere = `This happened while trying to create the plugin: ${params.name}`

  return {
    name: {
      constraint: params.name.startsWith('onda-plugin-'),
      message: `Plugin names should start with onda-runtime-. ${errorWhere}. Try onda-plugin-${params.name} instead`,
    },
    rules: {
      constraint:
        !params.rules ||
        Object.values(params.rules as Record<string, any>).every((rule) =>
          params.namespaces.includes(rule)
        ),
      message: `Rules can only access the namespaces you've declared. ${errorWhere}. Make sure that you're declaring all the namespaces the plugin will act in.`,
    },
  }
}

/**
 * Gets the plugin name
 * @param plugin
 */
export function getPluginName<Theme extends Record<string, any>>(
  plugin: Plugin<Theme>
) {
  return plugin[PLUGIN_NAME]
}

/**
 * Gets the plugin name
 * @param plugin
 */
export function getNamespaces<Theme extends Record<string, any>>(
  plugin: Plugin<Theme>
) {
  return plugin[PLUGIN_NAMESPACES]
}

/**
 * Gets the plugin entries
 * @param plugin
 */
export function getEntries<Theme extends Record<string, any>>(
  plugin: Plugin<Theme>
) {
  return plugin[PLUGIN_ENTRIES]
}

/**
 * Gets the plugin aliases
 * @param plugin
 */
export function getAliases<Theme extends Record<string, any>>(
  plugin: Plugin<Theme>
) {
  return plugin[PLUGIN_ALIASES]
}

/**
 * Gets the plugin rules
 * @param plugin
 */
export function getRules<Theme extends Record<string, any>>(
  plugin: Plugin<Theme>
) {
  return plugin[PLUGIN_RULES]
}

/**
 * Gets the plugin transforms
 * @param plugin
 */
export function getTransforms<Theme extends Record<string, any>>(
  plugin: Plugin<Theme>
) {
  return plugin[PLUGIN_TRANSFORMS]
}

/**
 * Gets the plugin splits
 * @param plugin
 */
export function getSplits<Theme extends Record<string, any>>(
  plugin: Plugin<Theme>
) {
  return plugin[PLUGIN_SPLITS]
}

/**
 * Check if the passed plugin is valid
 * @param plugin
 */
export function isValidPlugin<Theme extends Record<string, any>>(
  plugin: Plugin<Theme>
) {
  return (
    !!getPluginName(plugin) &&
    !!getNamespaces(plugin) &&
    !!getEntries(plugin) &&
    !!getAliases(plugin) &&
    !!getRules(plugin) &&
    !!getTransforms(plugin) &&
    !!getSplits(plugin)
  )
}

export interface Plugin<Theme extends Record<string, any>> {
  [PLUGIN_NAME]: string
  [PLUGIN_NAMESPACES]: string[]
  [PLUGIN_ENTRIES]: PluginOption<Theme, Partial<Theme>>
  [PLUGIN_ALIASES]: PluginOption<Theme, Record<string, string>>
  [PLUGIN_RULES]: PluginOption<Theme, Record<string, string>>
  [PLUGIN_TRANSFORMS]: PluginOption<
    Theme,
    Record<string, (rule: Rule, value: any) => any>
  >
  [PLUGIN_SPLITS]: PluginOption<Theme, Record<string, string[]>>
}

export type Rule = number[] | string[] | Record<string, any>

export interface PluginParams<Theme extends Record<string, any>> {
  name: string
  namespaces: string[]
  entries?: PluginOption<Theme, Partial<Theme>>
  aliases?: PluginOption<Theme, Record<string, string>>
  rules?: PluginOption<Theme, Record<string, string>>
  transforms?: PluginOption<
    Theme,
    Record<string, (rule: Rule, value: any) => any>
  >
  splits?: PluginOption<Theme, Record<string, string[]>>
}

export type PluginOption<Theme, ReturnType> =
  | ReturnType
  | ((t: Partial<Theme>) => ReturnType)

export interface StepsInstance {
  entries: {
    value: Record<string, any>
    exec: (theme: any) => any
  }
  aliases: {
    value: Record<string, string>
    exec: (prop: string) => string
  }
  rules: {
    value: Record<string, string>
    exec: (prop: string) => any
  }
  splits: {
    value: Record<string, string[]>
    exec: (prop: string, value: any) => Record<string, any>
  }
  transforms: (prop: string) => {
    value: Record<string, string>
    exec: (rule: Rule, value: any) => any
  }
}
