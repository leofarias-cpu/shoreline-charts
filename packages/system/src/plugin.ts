export function createPlugin<Theme extends Record<string, any>>(
  specs: PluginSpec<Theme>
): Plugin<Theme> {
  return {
    name: specs.name,
    onCreateAlias: specs?.onCreateAlias ?? noop,
    onCreateRule: specs?.onCreateRule ?? noop,
    onTransform: specs?.onTransform ?? noop,
    onSplit: specs?.onSplit ?? noop,
  }
}

export type Plugin<Theme extends Record<string, any>> = Required<
  PluginSpec<Theme>
>

export interface PluginSpec<Theme extends Record<string, any>> {
  name: string
  onCreateAlias?: (theme: Theme) => Record<string, string>
  onCreateRule?: (theme: Theme) => Record<string, string>
  onTransform?: (
    theme: Theme
  ) => Record<string, (rule: Record<string, string>, value: any) => any>
  onSplit?: (theme: Theme) => Record<string, string[]>
}

function noop() {
  return {}
}
