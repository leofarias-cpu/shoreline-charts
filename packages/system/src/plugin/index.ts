export * from './createPlugin'
export * from './steps'
export * from './types'
