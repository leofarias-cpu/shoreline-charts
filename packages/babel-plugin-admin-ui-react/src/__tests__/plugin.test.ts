import pluginTester from 'babel-plugin-tester'
import { adminUiPlugin } from '../plugin'

pluginTester({
  pluginName: '@vtex/babel-plugin-admin-ui-react',
  plugin: adminUiPlugin,
  snapshot: true,
  tests: {
    normal: `
      import { jsx } from '@vtex/admin-ui-react'

      const Box = jsx('div')({})
    `,
    'multiple calls': `
      import { jsx } from '@vtex/admin-ui-react'

      const Box = jsx('div')({})
      const Flex = jsx(Box)({ display: 'flex' })
      const Center = jsx(Flex)({ justifyContent: 'center' })
    `,
    'export const': `
      import { jsx } from '@vtex/admin-ui-react'

      export const Box = jsx('div')({})
    `,

    'has inner scope': `
      import { jsx } from '@vtex/admin-ui-react'

      const render = () => {
        const Box = jsx('div')({})
      }
    `,

    'shadowing imported jsx': `
      import { jsx } from '@vtex/admin-ui-react'

      const render = () => {
        const jsx = () => () => {}

        const Box = jsx('div')({})
      }
    `,

    'jsx of another module': `
      import { jsx } from 'another-module'

      const Box = jsx('div')({})
    `,
  },
})
