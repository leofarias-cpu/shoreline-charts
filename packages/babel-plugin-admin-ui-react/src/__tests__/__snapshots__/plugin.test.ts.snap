// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@vtex/babel-plugin-admin-ui-react export const: export const 1`] = `

import { jsx } from '@vtex/admin-ui-react'

export const Box = jsx('div')({})

      ↓ ↓ ↓ ↓ ↓ ↓

import { jsx } from '@vtex/admin-ui-react'
export const Box = Object.assign(jsx('div')({}), {
  displayName: 'Box',
})


`;

exports[`@vtex/babel-plugin-admin-ui-react has inner scope: has inner scope 1`] = `

import { jsx } from '@vtex/admin-ui-react'

const render = () => {
  const Box = jsx('div')({})
}

      ↓ ↓ ↓ ↓ ↓ ↓

import { jsx } from '@vtex/admin-ui-react'

const render = () => {
  const Box = Object.assign(jsx('div')({}), {
    displayName: 'Box',
  })
}


`;

exports[`@vtex/babel-plugin-admin-ui-react jsx of another module: jsx of another module 1`] = `

import { jsx } from 'another-module'

const Box = jsx('div')({})

      ↓ ↓ ↓ ↓ ↓ ↓

import { jsx } from 'another-module'
const Box = jsx('div')({})


`;

exports[`@vtex/babel-plugin-admin-ui-react multiple calls: multiple calls 1`] = `

import { jsx } from '@vtex/admin-ui-react'

const Box = jsx('div')({})
const Flex = jsx(Box)({ display: 'flex' })
const Center = jsx(Flex)({ justifyContent: 'center' })

      ↓ ↓ ↓ ↓ ↓ ↓

import { jsx } from '@vtex/admin-ui-react'
const Box = Object.assign(jsx('div')({}), {
  displayName: 'Box',
})
const Flex = Object.assign(
  jsx(Box)({
    display: 'flex',
  }),
  {
    displayName: 'Flex',
  }
)
const Center = Object.assign(
  jsx(Flex)({
    justifyContent: 'center',
  }),
  {
    displayName: 'Center',
  }
)


`;

exports[`@vtex/babel-plugin-admin-ui-react normal: normal 1`] = `

import { jsx } from '@vtex/admin-ui-react'

const Box = jsx('div')({})

      ↓ ↓ ↓ ↓ ↓ ↓

import { jsx } from '@vtex/admin-ui-react'
const Box = Object.assign(jsx('div')({}), {
  displayName: 'Box',
})


`;

exports[`@vtex/babel-plugin-admin-ui-react shadowing imported jsx: shadowing imported jsx 1`] = `

import { jsx } from '@vtex/admin-ui-react'

const render = () => {
  const jsx = () => () => {}

  const Box = jsx('div')({})
}

      ↓ ↓ ↓ ↓ ↓ ↓

import { jsx } from '@vtex/admin-ui-react'

const render = () => {
  const jsx = () => () => {}

  const Box = jsx('div')({})
}


`;
